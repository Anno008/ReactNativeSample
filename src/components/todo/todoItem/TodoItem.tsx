import React from "react";
import { Button, Text, TouchableOpacity } from "react-native";
import { View } from "react-native-animatable";
import { ITodoItem } from "../../../models/ITodoItem";
import styles from "./TodoItemStyle";

interface IProps {
  todo: ITodoItem;
  toggleTodo: (todoItem: ITodoItem) => void;
  deleteTodo: (id: number) => void;
}

export const TodoItem: React.SFC<IProps> = (props: IProps) => {
  const { todo, toggleTodo, deleteTodo } = props;
  const { finished, name, id } = todo;

  const strikeOutStyle = finished ? styles.strikeout : null;
  const toggleTodoHandler = () => toggleTodo(todo);
  const deleteTodoHandler = () => deleteTodo(id);

  return (
    <View animation="fadeInLeft">
      <TouchableOpacity style={styles.container} onPress={toggleTodoHandler}>
        <Text style={[styles.text, strikeOutStyle]}>{name}</Text>
        <Button title="Delete" onPress={deleteTodoHandler} />
      </TouchableOpacity>
    </View>
  );
};
