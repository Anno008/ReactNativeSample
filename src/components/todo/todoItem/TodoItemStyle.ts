import { StyleSheet } from "react-native";

export default StyleSheet.create({
  container: {
    alignItems: "center",
    alignSelf: "stretch",
    borderColor: "#d6d7da",
    borderRadius: 4,
    borderWidth: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    margin: 5,
    padding: 5
  },
  strikeout: {
    textDecorationLine: "line-through",
    textDecorationStyle: "solid",
  },
  text: {
    fontSize: 30,
    fontWeight: "bold"
  },
});
