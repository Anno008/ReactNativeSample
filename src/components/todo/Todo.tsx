import React, { Component } from "react";
import { ScrollView, Text, View } from "react-native";

import CreateTodo from "components/todo/createTodo";
import { TodoItem } from "components/todo/todoItem/TodoItem";
import styles from "components/todo/TodoStyle";
import { ITodoItem } from "models/ITodoItem";

interface Props {
    todos: ITodoItem[];
    toggleTodoItem: (id: number) => void;
    deleteTodoItem: (id: number) => void;
}

export class Todo extends Component<Props> {
    scrollView: React.RefObject<ScrollView>;

    constructor(props: Props) {
        super(props);
        this.scrollView = React.createRef<ScrollView>();
    }

    render(): JSX.Element {
        const { todos } = this.props;
        return (
            <View style={styles.container}>
                <Text style={styles.title}>Todos</Text>
                {this.renderTodos(todos)}
                <CreateTodo />
            </View>
        );
    }

    renderTodos = (todos: ITodoItem[]): JSX.Element => (
        <ScrollView
            contentContainerStyle={styles.scrollViewStyle}
            style={styles.todoContainer}
            ref={this.scrollView}
            onContentSizeChange={this.scrollToBottom}
        >
            {todos.map((todoItem: ITodoItem): JSX.Element => (
                <TodoItem
                    key={todoItem.id}
                    todo={todoItem}
                    toggleTodo={this.toggleTodo}
                    deleteTodo={this.deleteTodo}
                />
            ))}
        </ScrollView>
    )

    toggleTodo = (id: number): void => this.props.toggleTodoItem(id);

    deleteTodo = (id: number): void => this.props.deleteTodoItem(id);

    scrollToBottom = (): void => {
        if (this.scrollView.current) {
            this.scrollView.current.scrollToEnd();
        }
    }
}

export default Todo;
