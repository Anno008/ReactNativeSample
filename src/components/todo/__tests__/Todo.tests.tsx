import Todo from "components/todo/Todo";
import { TodoItem } from "components/todo/todoItem/TodoItem";
import { shallow, ShallowWrapper } from "enzyme";
import React from "react";

describe("Todo component test", () => {
  // Arrange on top as not to repeat it in every test case
  const mockTodoItem = {
    finished: false,
    id: 1,
    name: "test"
  };
  const mockProps = {
    deleteTodoItem: jest.fn(),
    todos: [mockTodoItem],
    toggleTodoItem: jest.fn()
  };

  it("should render", () => {
    const wrapper: ShallowWrapper = shallow(<Todo {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should call toggle todo on child action", () => {
    const wrapper: ShallowWrapper = shallow(<Todo {...mockProps} />);

    wrapper.find(TodoItem).prop("toggleTodo")(mockTodoItem.id);

    expect(mockProps.toggleTodoItem).toBeCalled();
    expect(mockProps.toggleTodoItem).toBeCalledWith(mockTodoItem.id);
  });

  it("should call delete todo on child action", () => {
    const wrapper: ShallowWrapper = shallow(<Todo {...mockProps} />);

    wrapper.find(TodoItem).prop("deleteTodo")(mockTodoItem.id);

    expect(mockProps.deleteTodoItem).toBeCalled();
    expect(mockProps.deleteTodoItem).toBeCalledWith(mockTodoItem.id);
  });
});
