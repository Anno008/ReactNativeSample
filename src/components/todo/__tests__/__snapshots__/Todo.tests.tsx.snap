// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component test should render 1`] = `ShallowWrapper {}`;
